core/src/Streamly/Internal/Data/Array.hs

```diff
@@ -306,30 +290,82 @@ strip p arr =
             let indexL = getIndexL 0 -- first predicate failing index
             in if indexL == 0 && indexR == lastIndex
                then arr
-               else cloneArray arr indexL (indexR - indexL + 1)
+               else unsafeFreeze $ unsafePerformIO $ do
+                   let newLen = indexR - indexL + 1
+                       arrThawed = unsafeThaw (sliceUnsafe indexL newLen arr)
+                   MArray.clone arrThawed
 
     where
 
     getIndexR idx
         | idx < 0 = idx
         | otherwise =
-            if p (indexArray arr idx) then getIndexR (idx - 1) else idx
+            if p (getIndexUnsafe arr idx) then getIndexR (idx - 1) else idx
 
-    getIndexL idx = if p (indexArray arr idx) then getIndexL (idx + 1) else idx
+    getIndexL idx = if p (getIndexUnsafe arr idx)
+                    then getIndexL (idx + 1)
+                    else idx
 
 -- | Write an input stream of (index, value) pairs to an array. Throws an
 -- error if any index is out of bounds.
 --
 -- /Pre-release/
 {-# INLINE putIndices #-}
-putIndices :: PrimMonad m
+putIndices :: MonadIO m
     => Array a -> Fold m (Int, a) ()
 putIndices arr = FL.rmapM (\ _ -> return ()) (FL.foldlM' step initial)
 
     where
 
-    initial =  unsafeThawArray arr
+    initial = return $ unsafeThaw arr
 
     step marr (i, x) = do
-        writeArray marr i x
+        MArray.putIndexUnsafe marr i x
         return marr
+
+-------------------------------------------------------------------------------
+-- Instances
+-------------------------------------------------------------------------------
+
+instance Eq a => Eq (Array a) where
+    a1 == a2 = lenA1 == lenA2 && loop (lenA1 - 1)
+
+        where
+
+        lenA1 = length a1
+        lenA2 = length a2
+
+        loop i
+            | i < 0 = True
+            | otherwise =
+                let v1 = getIndexUnsafe a1 i
+                    v2 = getIndexUnsafe a2 i
+                 in v1 == v2 && loop (i - 1)
+
+instance Ord a => Ord (Array a) where
+    compare a1 a2 =
+        case compare lenA1 lenA2 of
+            LT -> LT
+            GT -> GT
+            EQ -> loop 0
+
+        where
+
+        lenA1 = length a1
+        lenA2 = length a2
+
+        loop i
+            | i >= lenA1 = EQ
+            | otherwise =
+                let v1 = getIndexUnsafe a1 i
+                    v2 = getIndexUnsafe a2 i
+                 in case compare v1 v2 of
+                        LT -> LT
+                        GT -> GT
+                        EQ -> loop (i + 1)
+
+instance Show a => Show (Array a) where
+    show arr = show $ toList arr
```

Comment:
Should use something like "fromList" in the show output. Should be consistent with the unboxed array show/read instances.